### PR TITLE
Bump Ruby 3.0 image to 3.0.3, Ruby 3.1 image to 3.1.1, and add support for local testing/development on arm64

### PR DESCRIPTION
**Pull request by ivoanjo**
Title/individual commits say the whole story.

I also attempted to use the newly-created arm64 images on CircleCI... but they don't have arm64 support on their docker executor yet (<https://circleci.canny.io/cloud-feature-requests/p/support-arm-resource-class-on-docker-executor>).

I was able to do a full test suite run on arm64 locally (except for mongodb, presto and cucumber tests but I don't think those failures are related to arm64, will investigate later). 

It should be possible to add support for arm64 for older Rubies, but since it's quite a bit of manual work, I'll leave it for a later PR.

### CODE REVIEW

File: .circleci/images/primary/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Dockerfiles used for testing and development
+
+This folder contains Dockerfiles for all versions of Ruby used in the testing and development of dd-trace-rb.
+
+## Multiplatform builds
+
+Images marked with a "# This image supports multiple platforms" are able to be built for both amd64 (x86-64) and
```
Comment (lloeki) [inline .circleci/images/primary/README.md:7]:
```suggestion
Images marked with a "# This image supports multiple platforms" are able to be built for both amd64 (x86_64) and
```

Comment (ivoanjo) [reply .circleci/images/primary/README.md]: Applied, thanks 👍

File: Gemfile [hunk 2]
```diff
@@ -87,7 +87,7 @@ end
 # previously-passing codebase start failing. Thus, we need to lock to a specific
 # version and bump it from time to time.
 # Also, there's no support for windows
-if RUBY_VERSION >= '2.4.0' && !Gem.win_platform?
+if RUBY_VERSION >= '2.4.0' && !Gem.win_platform? && RUBY_PLATFORM != 'aarch64-linux'
```
Comment (lloeki) [inline Gemfile:90]: I think it makes sense to rather have this be `RUBY_PLATFORM =~ /^x86_64-(darwin|linux)/` instead, WDYT?

Angle is making it work out of the box for other platforms such as freebsd or ppc64le, and consistency with what `sorbet` is available on.

File: docker-compose.yml [hunk 3]
```diff
@@ -283,6 +283,34 @@ services:
       - .:/app
       - bundle-3.1:/usr/local/bundle
       - "ddagent_var_run:${TEST_DDAGENT_VAR_RUN}"
+  tracer-3.1-arm64:
```
Comment (lloeki) [inline docker-compose.yml:314]: `arm64` is not a good name, it doesn't even exist in ARM specs and is easily confused with `amd64`, it's only used by Darwin and qemu (and IIRC `aarch64` also works for qemu). I'd rather use `aarch64` which is in the ruby platform name, gcc target, and clang triplet.

**Review by lloeki — APPROVED** [review]
LGTM, a couple of arguable comments.

File: docker-compose.yml (hunk 3 above)
Comment (ivoanjo) [reply docker-compose.yml:314]: There's inconsistencies everywhere. Apple actually uses it on macOS: https://developer.apple.com/forums/thread/668206 .

If you feel strongly, I don't mind changing it, but I think "arm 64" is clearer than "aarch64".

File: Gemfile (hunk 2 above)
Comment (ivoanjo) [reply Gemfile]: Fixed in 249a803dd7b849b8bc94167add5fb5a1a4c875ae

**Comment by ivoanjo** [comment]
Rebased to fix conflict with master.

**Pull request merged**